vernac/lemmas.ml

```diff
@@ -178,56 +178,121 @@ let start_lemma_with_initialization ?hook ~poly ~scope ~kind ~udecl sigma recgua
 
-(* Helper for process_recthms *)
-let retrieve_first_recthm uctx = function
-  | GlobRef.VarRef id ->
-    NamedDecl.get_value (Global.lookup_named id),
-    Decls.variable_opacity id
-  | GlobRef.ConstRef cst ->
-    let cb = Global.lookup_constant cst in
-    (* we get the right order somehow but surely it could be enforced in a better way *)
-    let uctx = UState.context uctx in
-    let inst = Univ.UContext.instance uctx in
-    let map (c, _, _) = Vars.subst_instance_constr inst c in
-    (Option.map map (Global.body_of_constant_body Library.indirect_accessor cb), Declareops.is_opaque cb)
-  | _ -> assert false
-
-let rec select_body i t =
-  let open Constr in
-  match Constr.kind t with
-  | Fix ((nv,0),decls) -> mkFix ((nv,i),decls)
-  | CoFix (0,decls) -> mkCoFix (i,decls)
-  | LetIn(na,t1,ty,t2) -> mkLetIn (na,t1,ty, select_body i t2)
-  | Lambda(na,ty,t) -> mkLambda(na,ty, select_body i t)
-  | App (t, args) -> mkApp (select_body i t, args)
-  | _ ->
-    CErrors.anomaly
-      Pp.(str "Not a proof by induction: " ++
-          Termops.Internal.debug_print_constr (EConstr.of_constr t) ++ str ".")
-
-(* Helper for process_recthms *)
-let save_remaining_recthms ?fix_exn ~poly ~scope ~udecl ~hook uctx body opaque i { Recthm.name; typ; impargs } =
-  let sigma = Evd.from_ctx uctx in
-  let kind = Decls.(IsAssumption Conjectural) in
-  match body with
-  | None ->
-    let _, pe = DeclareDef.prepare_parameter ~allow_evars:false ~poly sigma udecl typ in
-    DeclareDef.declare_assumption ?fix_exn ~name ~scope ~hook ~impargs ~uctx pe
-  | Some body ->
-    let body = EConstr.of_constr (select_body i body) in
-    (* XXX: we are normalizing twice here, entries do contain ground terms *)
-    let _, de = DeclareDef.prepare_definition ~allow_evars:false ~opaque ~poly sigma udecl ~types:(Some typ) ~body in
-    let hook_data = Option.map (fun hook -> hook, uctx, []) hook in
-    let ubind = UnivNames.empty_binders in (* XXX fixme ubind *)
-    DeclareDef.declare_definition ~name ~scope ~kind ?hook_data ubind de impargs
-
-(* This declares implicits and calls the hooks for other_thms *)
-let process_recthms ?fix_exn ?hook uctx ~udecl ~poly ~scope dref other_thms =
-  if List.is_empty other_thms then ()
-  else
-    let body, opaque = retrieve_first_recthm uctx dref in
-    let _ = List.map_i (save_remaining_recthms ?fix_exn ~poly ~scope ~udecl ~hook uctx body opaque) 1 other_thms in
-    ()
+(* Support for mutually proved theorems *)
+
+(* XXX: Most of this does belong to Declare, due to proof_entry manip *)
+module MutualEntry : sig
+
+  (* We keep this type abstract and to avoid uncontrolled hacks *)
+  type t
+
+  val variable : other_thms:Recthm.t list -> Entries.parameter_entry -> t
+
+  val adjust_guardness_conditions
+    : other_thms:Recthm.t list
+    -> possible_indexes:int list list
+    -> Evd.side_effects Declare.proof_entry
+    -> t
+
+  val declare_mutind
```

Comment:
Why "mutind"?

Reply:
I meant "mutually inductive definition" but I agree that's a bad name, correcting.